### PR TITLE
find peptide formula and most abundant isotopic mass

### PR DESCRIPTION
**Pull request by kyp4**
this pr calculates a peptide's full chemical formula (including mods) and has the first draft of code to find the most abundant isotopic mass

### CODE REVIEW

File: FlashLFQ/Peptide.cs [hunk 1]
```diff
@@ -55,6 +56,11 @@ public double GetIntensity(SpectraFileInfo fileInfo)
             }
         }
 
+        public ChemicalFormula GetChemicalFormula()
```
Comment (trishorts) [inline FlashLFQ/Peptide.cs:59]: delete unused code

File: Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs [hunk 2]
```diff
@@ -110,6 +111,46 @@ public double MonoisotopicMass
                 return (double)ClassExtensions.RoundedDouble(_monoisotopicMass.Value);
             }
 
+        }
+        /*
+        public ChemicalFormula MostAbundantMass
```
Comment (trishorts) [inline Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs:116]: delete commented code or uncomment

File: Test/TestDeconvolution.cs [hunk 3]
```diff
@@ -15,6 +17,84 @@ namespace Test
     [ExcludeFromCodeCoverage]
     public sealed class TestDeconvolution
     {
+
+        [Test]
+        public static void CheckGetMostAbundantObservedIsotopicMass()
+        {
+            
+            string fullFilePathWithExtension = @"D:\TDBU\Jurkat\TD-Projects-JurkatTopDownSeanDaiPaper\FXN11_tr1_032017.raw"; //@"D:\Other\averageMass\for_shortreed_protein_mix_sid 15_topdown_standard-qb.raw";
```
Comment (trishorts) [inline Test/TestDeconvolution.cs:25]: can't hard code files paths. have to load unit test files into mzlib for storage. make sure the add "copy always" to the properties

File: Test/TestDeconvolution.cs [hunk 4]
```diff
@@ -15,6 +17,84 @@ namespace Test
...
+            ThermoRawFileReader staticRaw = ThermoRawFileReader.LoadAllStaticData(fullFilePathWithExtension);
+            List<MsDataScan> scan = staticRaw.GetAllScansList(); //.Where(p => p.RetentionTime == 35.88835).ToList();
+            scan = scan.Where(p => p.MsnOrder == 1).ToList();
+            scan = scan.Where(p => p.OneBasedScanNumber == 587).ToList();
+
+            //scan = scan.Where(p => p.RetentionTime < 35.888).Where(p => p.RetentionTime > 35.8).ToList();
```
Comment (trishorts) [inline Test/TestDeconvolution.cs:31]: delete comment

File: Test/TestDeconvolution.cs [hunk 5]
```diff
@@ -15,6 +17,84 @@ namespace Test
...
+            //List<MassSpectrometry.IsotopicEnvelope> lie = scan[3000].GetIsolatedMassesAndCharges(spec, 1, 60, 4, 3).ToList();
+            lie = lie.Where(p => p.MostAbundantObservedIsotopicMass < 8246.5+1).OrderByDescending(p => p.MostAbundantObservedIsotopicMass).ToList();
+            var compare = lie.Select(p => p.MostAbundantObservedIsotopicMass).ToList()[0];
+
+            Assert.That(compare, Is.EqualTo(8246.5).Within(1));
+
+
+            //List<IsotopicEnvelope>
+
+            //List<MassSpectrometry.IsotopicEnvelope> lie = spec.Deconvolute(theRange, minAssumedChargeState, maxAssumedChargeState, deconvolutionTolerancePpm, intensityRatioLimit).ToList();
```
Comment (trishorts) [inline Test/TestDeconvolution.cs:52]: delete

File: Test/TestPeptideWithSetMods.cs [hunk 6]
```diff
@@ -929,5 +930,60 @@ public static void CountTargetsWithMatchingDecoys()
...
+            PeptideWithSetModifications large_pep_mod = new PeptideWithSetModifications(new Protein("PEPSIDEKRNSPEPTIDEKECUEIRQUV", "ASSESSION"), new DigestionParams(protease: "trypsin"), 1, 28, CleavageSpecificity.Full, null, 0, modDict_large, 0, null);
+            ChemicalFormula large_pep_mod_cf = ChemicalFormula.ParseFormula("C135H223N38O57P2S1Se2");
+            Assert.AreEqual(large_pep_mod.FullChemicalFormula, large_pep_mod_cf);
+        }
+
+
+        /*
+        [Test]
+        public static void CheckMostAbundantMonoIsotopicMass()
+        {
```
Comment (trishorts) [inline Test/TestPeptideWithSetMods.cs:971]: comment

File: MassSpectrometry/MzSpectra/IsotopicEnvelope.cs [hunk 7]
```diff
@@ -1,30 +1,39 @@
...
             Charge = bestChargeState;
             TotalIntensity = bestTotalIntensity;
             StDev = bestStDev;
             MassIndex = bestMassIndex;
             Score = ScoreIsotopeEnvelope();
         }
 
+        public double GetMostAbundantObservedIsotopicMass(List<(double mz, double intensity)>  peaks, int charge)
+        {
+            return peaks.OrderByDescending(p => p.intensity).ToList()[0].Item1 * charge;
```
Comment (avcarr2) [inline MassSpectrometry/MzSpectra/IsotopicEnvelope.cs:34]: I'm not sure about this formula. I think formula for most abundant isotopic mass (IM) of a given charge state (z) should be (IM - z * hydrogen mass) / z.

**Review by avcarr2 — CHANGES_REQUESTED** [review]
See my comment on the most isotopic mass identification

**Pull request closed without merging**